### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive for `do { ... } while (false)` when allowConstantLoopConditions is true

### ISSUE THREAD

**Issue by justinyaodu** [typescript-eslint/typescript-eslint#7688]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=CYewBA3gUGYJYDMwAoCyBDALgCwHQCd0A7UAW2QEowAeMABlwFYrpZYAjfAU3QGsBuGGAC%2BQgMYgiAZxAAbLrlkgA5sgBEWMPPRTM9JmoqDhYAO7Y48lAnSypXI1CA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2uGw5E0aB2iQANKrVY12SEPjwOAdwDCk5PiEsAMhw7Fb02fKaoM%2BOIgG2AC%2BQQC6qqHBQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoAkNZOgB4CG0euBxTqACprIljwBPACpCBZSnkwBXdABpwEKCWnJWAORmpUAYQAW6aAGtJYaXKUBfEFaA&tokens=false

### Repro Code

```typescript
do {
  if (Math.random() < 0.5) {
    break;
  }
  console.log("at least 0.5");
} while (false);
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": [
      "error",
      {
        "allowConstantLoopConditions": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "exactOptionalPropertyTypes": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected no errors to be reported, because `false` is a constant loop condition, and I have `allowConstantLoopConditions` enabled.

`do { ... } while (false)` is useful for creating a block that runs once, while supporting `break` statements to exit the block early.

### Actual Result

On line 6, I get the error "Unnecessary conditional, value is always falsy."

### Additional Info

If we agree that this behavior should be changed, I would be happy to work on a patch for this.

**Comment by bradzacher** [comment]
The option is actually hard coded to specifically look for `true` only.
TBH I've never ever heard of people using this pattern! It's literally the first I've ever seen it. It's likely not a common pattern either given this is the first issue we've gotten about it.

The ways I've seen of doing this are:

```TS
// IIFE
(() => {
  if (Math.random() < 0.5) {
    break;
  }
  console.log("at least 0.5"); 
})();

// or even more common and idiomatic...
// if..else
if (Math.random() < 0.5) {
  break;
} else {
  console.log("at least 0.5");
}
```

TBH I'm not sure this is a special case we would want to support given how rare it is.
If I saw this in regular code I'd personally first assume that it was a bug and wonder why the rule didn't error.

WDYT @typescript-eslint/triage-team

**Comment by JamesHenry** [comment]
I’m inclined to agree. If I saw the code form the OP in a PR review I would almost certainly request it be simplified to aid with future readability.

**Comment by JamesHenry** [comment]
But then again maybe that’s overstepping? The lint rule claims to do something, and it doesn’t do it fully… maybe it’s not up to us to police funky code patterns indirectly via lint rules like that? It wouldn’t be expensive to check for false too right?

**Comment by JamesHenry** [comment]
It’s genuinely an interesting case of where we draw certain lines

**Comment by justinyaodu** [comment]
After reading your comments and searching up `do { ... } while (false)` online, I've realized that this is less common and more controversial than I expected:

Here are some StackOverflow posts regarding this code pattern in [C++](https://stackoverflow.com/questions/2314066/do-whilefalse) and [Java](https://stackoverflow.com/questions/1445025/what-is-the-advantage-of-a-do-whilefalse).

Meanwhile, the first relevant search result for JavaScript is a [Nim forum post](https://forum.nim-lang.org/t/8105) asking why `do { ... } while (false)` appears in the compiler output when compiling Nim to JavaScript. It seems like humans don't write this kind of JavaScript often!

To play devil's advocate, `do { ... } while (false)` is most useful in the following niche use case (which is why I tried it in the first place):
1. A block of code needs to run exactly once
2. Multiple early exit points, which make nested `if`-statements unwieldy
3. Exceptions occurring within the block should not be caught, ruling out `try { ... throw ... } catch {}`
4. Performance-sensitive, which made me hesitate to use an IIFE
5. JavaScript doesn't support `goto`, which is probably for the better in general, though I have seen C code which uses `goto` instead of `do { ... } while (false)`

However, I would totally understand if you decide to not support this. I'm okay with disabling no-unnecessary-condition if I want to write this kind of questionable code :sweat_smile:

**Comment by justinyaodu** [comment]
Here is a slightly more realistic repro (though with something this small, if-statements would still be more readable):

```js
do {
  const value = Math.random();
    if (value < 0.5) break;
    console.log("at least 0.5");
    if (value < 0.75) break;
    console.log("at least 0.75");
} while (false); 
```

I found that this can also be written using labels:

```js
test: {
  const value = Math.random();
    if (value < 0.5) break test;
    console.log("at least 0.5");
    if (value < 0.75) break test;
    console.log("at least 0.75");
}
```

**Comment by Josh-Cena** [comment]
Was about to say you can use labels with `break`, if you don't hate those.

**Comment by justinyaodu** [comment]
After thinking about this some more, even in that niche use case, I haven't found a situation where `do { ... } while (false)` is preferable to `label: { ... break label; ... }`.

Although the label syntax is rarely seen, I also think it would be less confusing overall. It's harder to tell at a glance whether `do { ... } while (false)` is an actual loop, or a block that is only executed once.

It sounds like there isn't sufficient motivation to change the current behavior, so I'll close the issue for now. Thanks for the great discussion, and for an excellent linting experience overall!

**Comment by JoshuaKGoldberg** [comment]
FWIW this brought me joy to read 😄 thanks for posting & discussing!

**Comment by JoshuaKGoldberg** [comment]
FWIW this thread brought me joy to read 😄 thanks for posting & discussing!
